### PR TITLE
Added amp-user-notification with a backend service

### PR DESCRIPTION
**Pull request by kul3r4**
Added amp-user-notification with a backend service called by the parameter data-show-if-href and data-dismissed-if-href.
Added a comment on the readme pointing at the project amp-publisher-sample that can be used as a local server to demo this feature.

### CODE REVIEW

**Comment by kul3r4** [comment]
I signed it!

2016-02-18 20:39 GMT+00:00 googlebot <notifications@github.com>:

> Thanks for your pull request. It looks like this may be your first
> contribution to a Google open source project. Before we can look at your
> pull request, you'll need to sign a Contributor License Agreement (CLA).
>
> [image: :memo:] *Please visit https://cla.developers.google.com/
> <https://cla.developers.google.com/> to sign.*
>
> Once you've signed, please reply here (e.g. I signed it!) and we'll
> verify. Thanks.
> ------------------------------
>
>    - If you've already signed a CLA, it's possible we don't have your
>    GitHub username or you're using a different email address. Check your
>    existing CLA data <https://cla.developers.google.com/clas> and verify
>    that your email is set on your git commits
>    <https://help.github.com/articles/setting-your-email-in-git/>.
>    - If you signed the CLA as a corporation, please let us know the
>    company's name.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ampproject/amp-by-example/pull/4#issuecomment-185908228>
> .
>



-- 
Chiara Chiappini

File: README.md [hunk 1]
```diff
@@ -45,6 +45,11 @@ $ gulp
 $ open http://localhost:8000/index.html
 ```
 
+Some components, like the [amp-user-notification(with backend service)](https://amp-by-example.appspot.com/amp-user-notification(with_backend_service).html) require a server to
+run.
+
+If you want to run the demo locally, you can use the server from [the amp-publisher-sample](https://github.com/ampproject/amp-publisher-sample) project.
+
```
Comment (sebastianbenz) [inline README.md]: This won't work as the samples don't link to localhost. I don't think that we need this here.

File: src/amp-user-notification(with_backend_service).html [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    }
+  </style>
+  <style>body {opacity: 0}</style><noscript><style>body {opacity: 1}</style></noscript>
+  <script async src="https://cdn.ampproject.org/v0.js"></script>
+</head>
+<body>
+
+  <!-- #### Using Backend service -->
+  <!--
+    When using a backend solution, 2 parameters are available: data-show-if-href and
```
Comment (sebastianbenz) [inline src/amp-user-notification(with_backend_service).html]: replace: '2' with 'two'
mark parameters as code: `data-show-if-href` and `data-dismiss-if-href`

File: src/amp-user-notification(with_backend_service).html [hunk 3]
```diff
@@ -0,0 +1,80 @@
+<!--
```
Comment (sebastianbenz) [inline src/amp-user-notification(with_backend_service).html]: Could we put this in the same sample and show both dialogs above each other? Each one could say which version it is.

File: src/amp-user-notification(with_backend_service).html [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+  <meta charset="utf-8">
+  <meta name="viewport" content="width=device-width,minimum-scale=1,initial-scale=1">
+  <!-- #### Setup -->
+  <!-- Import the amp-user-notification component in the header. -->
+  <script async custom-element="amp-user-notification" src="https://cdn.ampproject.org/v0/amp-user-notification-0.1.js"></script>
+  <!--
+    Specify any additional css, default css class will put the amp-user-notification element
+    in the left bottom corner.
+  -->
+  <style amp-custom>
```
Comment (sebastianbenz) [inline src/amp-user-notification(with_backend_service).html]: can we keep the css a bit simpler like in the other sample?

File: src/amp-user-notification(with_backend_service).html [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+<body>
+
+  <!-- #### Using Backend service -->
+  <!--
+    When using a backend solution, 2 parameters are available: data-show-if-href and
+    data-dismiss-if-href. They can be used together or separatively.
+
+    The URL defined by data-show-if-href url will be queried once and you can use this to 
+    make a decision. For example you can use it to decide if showing a notification 
+    based on some user attribute.
```
Comment (sebastianbenz) [inline src/amp-user-notification(with_backend_service).html]: Can you make this more explicit?

File: src/amp-user-notification(with_backend_service).html [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+    The URL defined by data-dismiss-href will be called when the notification is dismissed 
+    and you can use it to make recordings on your side.
```
Comment (sebastianbenz) [inline src/amp-user-notification(with_backend_service).html]: What does recording mean?

File: src/amp-user-notification(with_backend_service).html [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+    If the notification has already been dismissed and the related item has been added to
+    the localstorage, the notification will not appear even if you have configured these 
+    parameters.
```
Comment (sebastianbenz) [inline src/amp-user-notification(with_backend_service).html]: which parameters?

**Comment by sebastianbenz** [comment]
Thanks! I have a few comments.

File: src/amp-user-notification(with_backend_service).html (hunk 3 above)
Comment (kul3r4) [inline src/amp-user-notification(with_backend_service).html]: This is not possible as explained here: https://github.com/ampproject/amphtml/blob/master/extensions/amp-user-notification/amp-user-notification.md

File: src/amp-user-notification(with_backend_service).html (hunk 4 above)
Comment (kul3r4) [inline src/amp-user-notification(with_backend_service).html]: Agreed.

File: src/amp-user-notification(with_backend_service).html (hunk 2 above)
Comment (kul3r4) [inline src/amp-user-notification(with_backend_service).html]: Ok.

File: README.md [hunk 8]
```diff
@@ -45,6 +45,11 @@ $ gulp
 $ open http://localhost:8000/index.html
 ```
 
+Some components, like the [amp-user-notification(with backend service)](https://amp-by-example.appspot.com/amp-user-notification(with_backend_service).html) require a server to
+run.
```
Comment (sebastianbenz) [inline README.md]: replace with: an additional server endpoint.

File: src/amp-user-notification(with_backend_service).html (hunk 3 above)
Comment (sebastianbenz) [inline src/amp-user-notification(with_backend_service).html]: OK - let's rename then both samples (just rename the files), e.g.:

* 'amp-user-notification_with_local_storage'
* 'amp-user-notification_with_server_endpoint'

File: src/amp-user-notification(with_backend_service).html (hunk 5 above)
Comment (sebastianbenz) [inline src/amp-user-notification(with_backend_service).html]: e.g.: `data-show-if-href` needs to point to a server-endpoint returning `{ "showNotification": true|false }`. You can find a sample implementation [here](https://github.com/ampproject/amp-publisher-sample/blob/master/controllers/amp-user-notification/api.js).

File: src/amp-user-notification_with_local_storage.html [hunk 9]
```diff
@@ -33,12 +34,12 @@
   <!-- #### Using Local Storage -->
   <!--
     When using the local storage solution, the parameters data-show-if-href and
-    data-dismiss-if-href are optional. When the user agrees to the cookies policy,
-    a new item is added to the local storage.
+    data-dismiss-if-href are optional.
```
Comment (sebastianbenz) [inline src/amp-user-notification_with_local_storage.html]: Remove this sentence. If users read this article first,  they don't know what these parameters are.

File: src/amp-user-notification_with_local_storage.html [hunk 10]
```diff
@@ -33,12 +34,12 @@
   <!-- #### Using Local Storage -->
   <!--
     When using the local storage solution, the parameters data-show-if-href and
-    data-dismiss-if-href are optional. When the user agrees to the cookies policy,
-    a new item is added to the local storage.
+    data-dismiss-if-href are optional. 
+    When the user agrees to the cookies policy, a new item is added to the local storage.
```
Comment (sebastianbenz) [inline src/amp-user-notification_with_local_storage.html]: s/a new item is added/the status is saved/g

File: src/amp-user-notification_with_local_storage.html [hunk 11]
```diff
@@ -2,7 +2,7 @@
   #### Introduction
 
   Use amp-user-notification to display a dismissable notification to the user.
```
Comment (sebastianbenz) [inline src/amp-user-notification_with_local_storage.html]: wrap amp-user-notification in `

File: src/amp-user-notification_with_server_endpoint.html [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+<html ⚡>
+<head>
+  <meta charset="utf-8">
+  <meta name="viewport" content="width=device-width,minimum-scale=1,initial-scale=1">
+  <!-- #### Setup -->
+  <!-- Import the amp-user-notification component in the header. -->
+  <script async custom-element="amp-user-notification" src="https://cdn.ampproject.org/v0/amp-user-notification-0.1.js"></script>
+  <!--
+    Specify any additional css, default css class will put the amp-user-notification element
+    in the left bottom corner.
```
Comment (sebastianbenz) [inline src/amp-user-notification_with_server_endpoint.html]: replace 'default css class will put the amp-user-notification element in the left bottom corner' with 'by default, the `amp-user-notification` is positioned in the left bottom corner.

File: src/amp-user-notification_with_server_endpoint.html [hunk 13]
```diff
@@ -0,0 +1,59 @@
+<!--
+  #### Introduction
+
+  Use amp-user-notification to display a dismissable notification to the user.
```
Comment (sebastianbenz) [inline src/amp-user-notification_with_server_endpoint.html]: Can we add a sentence about when to use server-side storage, e.g. 'use user-notifications together with a server endpoint if you want to store the dismissal state across your AMP articles and your normal articles.'

**Comment by sebastianbenz** [comment]
Looks good - some comments regarding the docs.

File: src/amp-user-notification_with_server_endpoint.html (hunk 13 above)
Comment (kul3r4) [inline src/amp-user-notification_with_server_endpoint.html]: PTAL

**Comment by kul3r4** [comment]
PTAL

File: src/amp-user-notification_with_server_endpoint.html [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+</head>
+<body>
+
+  <!-- #### Using Backend service -->
+  <!--
+    When using a backend solution, two parameters are available: `data-show-if-href` and
+    `data-dismiss-if-href`. They can be used together or separatively.
+
+    `data-show-if-href` needs to point to a server-endpoint returning 
+    { "showNotification": true|false }.
```
Comment (sebastianbenz) [inline src/amp-user-notification_with_server_endpoint.html]: sorry one more thing: wrap { "showNotification": true|false } in `

File: src/amp-user-notification_with_server_endpoint.html [hunk 15]
```diff
@@ -0,0 +1,60 @@
+<!--
+  #### Introduction
+
+  Use `amp-user-notification` to display a dismissable notification to the user.
+  Use `user-notifications` together with a server endpoint if you want to store the dismissal
```
Comment (sebastianbenz) [inline src/amp-user-notification_with_server_endpoint.html]: even better, replace 'Use `user-notifications` together' with 'Combine it '

**Comment by kul3r4** [comment]
PTAL

**Comment by sebastianbenz** [comment]
Thanks!

**Pull request merged**